text/0000-externally-implementable-functions.md

```diff
@@ -0,0 +1,179 @@
+
+*Implementing* the function can be done by any crate that can name the item.
+(The `impl fn` item will need to name the item to implement, which could be directly or through an alias/re-export.)
+
+# Implementation
+
+The implementation will be based on the same mechanisms as used today for the `panic_handler` and `#[global_allocator]` features.
+
+The compiler of the root crate will find the implementation of all externally implementable functions and give an error
+if more than one implementation is found for any of them.
+If none are found, the result is either an error, or—if the `extern impl fn` has a default body—an implementation
+is generated that calls that default body.
+
+# Drawbacks
+
+- It encourages globally defined behaviour.
+  - Counterargument: We are already doing this anyway, both inside the standard library (e.g. panic_handler, allocator)
+    and outside (e.g. global logger). This just makes it much easier (and safer) to get right.
+- This will invite the addition of many hooks to the standard library to modify existing behavior.
+  While we should consider such possibilities, this RFC does not propose that every piece of standard library behavior should be replaceable.
+
+# Rationale and alternatives
+
+## Syntax
+
+The syntax re-uses existing keywords. Alternatively, we could:
+  - Use the `override` reserved keyword.
+  - Add a new (contextual) keyword (e.g. `existential fn`).
+  - Use an attribute (e.g. `#[extern_impl]`) instead.
+
+## Functions or statics
+
+This RFC only proposes externally implementable *functions*.
+
+An alternative is to only provide externally definable *statics* instead.
+
+That would be equivalent in power: one can store a function pointer in a static, and one can return a reference to a static from a function ([RFC 3635](https://github.com/rust-lang/rfcs/pull/3635)).
+
+(Another alternative, of course is to provide both. See future possibilities.)
+
+## Visibility
+
+There are two kinds of visibilities to be considered for externally implementable functions:
+who can *implement* the function, and who can *call* the function.
+
+Not allowing the function to be implemented by other crates nullifies the functionality, as the entire point of externally implementable functions is that they can be implemented in another crate. This visibility is therefore always (implicitly) "pub".
+
+Allowing a more restricted (that is, not `pub`) visibility for *calling* the function can be useful. For example, today's `#[panic_handler]` can be defined by any crate, but can not be called directly. (Only indirectly through `panic!()` and friends.)
+
+A downside is that it is not possible to allow this "only implementable but not publicly callable" visibility through an alias.
+
+An alternative could be to use the same visibility for both implementing an calling, which would simply mean that the function (or an alias to it) will always have to be `pub`.
+
```

Comment:
```suggestion

## Configuration

An `extern impl fn` may have a `#[cfg(...)]` applied to it as usual. For instance, a crate may only provide an `extern impl fn` with a given feature flag enabled, and might then use the same feature flag to conditionally provide make other functions depending on that `extern impl fn`. This is a useful pattern for crates that don't want to provide a default implementation but want to avoid producing a compilation error unless the function is needed.

## Stability

`extern impl fn` may have stability markers (e.g. `#[unstable]`) applied to it, in which case implementing the function requires opting into the unstable feature (and a compiler that permits opting into unstable features).

```

Reply:
Not sure if this is a useful addition to the RFC.

- cfg() doesn't only apply to `extern impl fn` items, also to `impl fn` items. But since it just works the same for all items, this doesn't really seem to add anything new.

- Stability attributes are unstable, so how that works is more of an implementation detail than part of the language.

Reply:
@m-ou-se I realize that configuration and stability attributes apply to just about everything. But it seems particularly relevant to observe that configuration attributes apply, because that serves as a viable substitute in lieu of handling a missing implementation when an externally implementable function is never called. I wrote this patch as an attempt to resolve the "statically-unused-extern-impl-fn" concern.

(That purpose is served entirely by the "Configuration" section, and I agree that the "Stability" section is less important since it's documenting the interaction with an unstable mechanism.)

Reply:
Alright. Added the note on configuration.